File: 0003-longest-substring-without-repeating-characters-无重复字符的最长子串/0003.cpp
// int [26] 用于字母 ‘a’ - ‘z’ 或 ‘A’ - ‘Z’
// int [128] 用于ASCII码
// int [256] 用于扩展ASCII码

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int len = 0;
        auto beg = -1;
        vector<int> table(128, -1);

        for (int p = 0; p != s.size(); ++p) {
            beg = max(table[s[p]], beg);
            table[s[p]] = p;
            len = max(p - beg, len);
        }

        return len;
    }
};

File: 0003-longest-substring-without-repeating-characters-无重复字符的最长子串/0003.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if len(s) == 0: return 0
        beg = 0
        end = 0
        length = 1
        table = {s[0]: 0}
        while end != len(s) - 1:
            end += 1
            if s[end] in table.keys():
                beg = max(table[s[end]] + 1, beg)
            table[s[end]] = end
            length = max(end - beg + 1, length)
        
        return length